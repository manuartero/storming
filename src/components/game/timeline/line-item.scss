@use '/src/styles/colors.scss';

$line-item-size: 42px;
$content-size: 20px;

.line-item {
  position: relative;
  top: 35px;

  display: flex;
  justify-content: center;
  align-items: center;

  width: $line-item-size;
  height: $line-item-size;
  border-radius: 50%;
  border: 2px solid black;
  background-color: colors.$dark-grey;



  &__content {
    width: $content-size;
    height: $content-size;
    border-radius: 50%;

    &--player {
      background-color: colors.$player-highlight;
    }

    // FIXME: move :hover{} to .line-item (instead of .line-item__content)
    &--player:hover {
      width: 100%;
      height: 100%;

      // TODO change background image to card icon
    }

    &--enemy1 {
      background-color: colors.$enemy1-highlight;
    }

    &--enemy2 {
      background-color: colors.$enemy2-highlight;
    }

    &--enemy3 {
      background-color: colors.$enemy3-highlight;
    }

    transition: width 0.25s ease-out,
    height 0.25s ease-out;
  }
}
