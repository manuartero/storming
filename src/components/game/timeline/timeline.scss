@use '/src/styles/colors.scss';
@use '/src/styles/fonts.scss';

.timeline {
  height: 92px;

  display: flex;
  flex-direction: row;
  flex-wrap: nowrap; // single line

  &__next {
    flex-grow: 3;
    background-color: colors.$base-grey;
  }

  &__future {
    flex-grow: 2;
    background-color: colors.$dark-grey;
  }

  &__section {
    position: relative;
    display: flex;
    padding: 0 20px 0 80px;
    border-radius: 0px 0px 12px 12px;

    &__name {
      position: absolute;
      top: 24px;
      left: 12px;

      @include fonts.heading();
    }

    &__line {
      width: 100%;
      position: relative;

      display: flex;
      flex-direction: row;
      justify-content: space-around;

      // black line
      &::before {
        content: '';
        height: 0px;
        width: 100%;

        position: absolute;
        top: 55px; // from timeline__section

        border-top: 2px solid black;
      }
    }
  }

}
