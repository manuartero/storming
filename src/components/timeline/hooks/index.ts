import { useReducer } from "react";
import initialTimeline from "./initial-timeline";
import timelineReducer from "./timeline-reducer";

export function useTimeline() {
  
  const [timelineState, dispatchTimelineAction] = useReducer(
    timelineReducer,
    initialTimeline
  );

  return { timelineState, dispatchTimelineAction };
}
