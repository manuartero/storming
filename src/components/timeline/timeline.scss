@use "/src/styles/fonts.scss";

.timeline {
  height: 13vh;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap; // single line

  &__next {
    flex: 1;
    background-color: var(--base-grey);
  }

  &__future {
    flex: 1;
    background-color: var(--dark-grey);
  }

  &__section {
    position: relative;
    display: flex;
    padding: 0 20px;
    border-radius: 0px 0px 12px 12px;

    &:hover {
      .timeline__section__name {
        opacity: 0.4;
        transition: opacity 0.5s ease-in-out;
      }
    }

    &__name {
      position: absolute; // relative to timeline__section
      top: 18px;
      left: 12px;
      opacity: 0;
      @include fonts.heading();
    }

    &__line {
      width: 100%;
      position: relative;

      display: flex;
      flex-direction: row;
      justify-content: space-around;

      // black line
      &::before {
        content: "";
        height: 0px;
        width: 100%;

        position: absolute;
        top: 50%;
        transform: translateY(-50%);

        border-top: 2px solid var(--darkest-grey);
      }
    }
  }
}
