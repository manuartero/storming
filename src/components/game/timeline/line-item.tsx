import c from "classnames";

import "./line-item.scss";

interface ActionLineItem {
  card: ActionCard;
}

export function ActionLineItem({ card }: ActionLineItem): JSX.Element {
  return (
    <div className={c("line-item", card.owner === "player" && "expansible")}>
      <div
        className={c("line-item__content", `line-item__content--${card.owner}`)}
      />
    </div>
  );
}
