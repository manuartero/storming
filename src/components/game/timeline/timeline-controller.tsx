import { useGameContext } from "contexts";
import { Timeline } from "./timeline";

/**
 * Renders depends on:
 *  - useGameContext()
 */
export function TimelineController(): JSX.Element {
  const gameContext = useGameContext();

  return <Timeline timeline={gameContext.timeline} />;
}
