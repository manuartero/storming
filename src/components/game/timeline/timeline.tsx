import c from "classnames";
import { logRender } from "utils/console";

import "./timeline.scss";

interface Props {
  timeline: Timeline;
}

function Timeline({ timeline }: Props): JSX.Element {
  logRender("Timeline");
  return (
    <div className="timeline">
      <div className="timeline__next timeline__section">
        <span className="timeline__section__name">NEXT</span>
        <div className="timeline__section__line">
          {timeline.next.map((card, i) => (
            <LineItem key={`timeline-next-item-${i}`} card={card} />
          ))}
        </div>
      </div>
      <div className="timeline__future timeline__section">
        <span className="timeline__section__name">FUTURE</span>
        <div className="timeline__section__line">
          {timeline.future.map((card, i) => (
            <LineItem key={`timeline-future-item-${i}`} card={card} />
          ))}
        </div>
      </div>
    </div>
  );
}

interface LineItemProps {
  card: Card;
}

function LineItem({ card }: LineItemProps): JSX.Element {
  // TODO on hover: show item
  const wasPlayedByTheUser =
    card.cardType === "actionCard" && card.owner === "player";

  return (
    <>
      <span
        className={c(
          "line-item",
          card.cardType === "actionCard" && `line-item--${card.owner}`
        )}
      />
      {wasPlayedByTheUser && <div className="line-item__detail"></div>}
    </>
  );
}

export default Timeline;
