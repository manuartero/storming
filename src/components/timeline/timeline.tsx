import c from "classnames";
import { LayoutGroup, motion } from "framer-motion";
import { ActionLineItem } from "./line-item";

import styles from "./timeline.module.css";

type Props = {
  next: TimelineCard[];
  future: TimelineCard[];
};

export function Timeline({ next, future }: Props) {
  const renderLineItems = (section: TimelineCard[]) => {
    return section.map(({ card, commited }) => {
      if (card.cardType === "actionCard")
        return (
          <motion.div
            key={card.cardId}
            layout
            initial={{ opacity: 0, scale: 0 }}
            animate={{
              opacity: [0, 0.9, 0.9, 1, 1, 1, 1],
              scale: [0, 1.3, 0.8, 1.2, 0.9, 1.1, 1],
            }}
            exit={{ opacity: 0, scale: 0 }}
            transition={{ duration: 0.3 }}
          >
            <ActionLineItem card={card} commited={commited} />
          </motion.div>
        );
    });
  };

  return (
    <div className={styles.timeline}>
      <div className={c(styles.next, styles.section)}>
        <span className={styles.sectionName}>NEXT</span>
        <div className={styles.line}>
          <LayoutGroup id="next-timeline">{renderLineItems(next)}</LayoutGroup>
        </div>
      </div>
      <div className={c(styles.future, styles.section)}>
        <span className={styles.sectionName}>FUTURE</span>
        <div className={styles.line}>
          <LayoutGroup id="future-timeline">
            {renderLineItems(future)}
          </LayoutGroup>
        </div>
      </div>
    </div>
  );
}
