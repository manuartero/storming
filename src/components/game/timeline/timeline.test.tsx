import { render } from "@testing-library/react";
import { Timeline } from "./timeline";

describe("<Timeline />", () => {
  test("render: match snapshot", () => {
    const { asFragment } = render(<Timeline timeline={timeline} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

const timeline: Timeline = {
  current: {
    cardType: "actionCard",
    action: "move",
    owner: "player",
    cardId: "player_move_A",
  },
  next: [
    {
      cardType: "actionCard",
      action: "move",
      owner: "enemy1",
      cardId: "enemy1_move_A",
    },
    {
      cardType: "actionCard",
      action: "move",
      owner: "enemy2",
      cardId: "enemy2_move_A",
    },
    {
      cardType: "actionCard",
      action: "move",
      owner: "enemy3",
      cardId: "enemy3_move_A",
    },
  ],
  future: [],
};
