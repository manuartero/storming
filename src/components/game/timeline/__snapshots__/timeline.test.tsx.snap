// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timeline /> render: match snapshot 1`] = `
<DocumentFragment>
  <div
    class="timeline"
  >
    <div
      class="timeline__next timeline__section"
    >
      <span
        class="timeline__section__name"
      >
        NEXT
      </span>
      <div
        class="timeline__section__line"
      >
        <div
          class="line-item"
        >
          <div
            class="line-item__content line-item__content--enemy1"
          />
        </div>
        <div
          class="line-item"
        >
          <div
            class="line-item__content line-item__content--enemy2"
          />
        </div>
        <div
          class="line-item"
        >
          <div
            class="line-item__content line-item__content--enemy3"
          />
        </div>
      </div>
    </div>
    <div
      class="timeline__future timeline__section"
    >
      <span
        class="timeline__section__name"
      >
        FUTURE
      </span>
      <div
        class="timeline__section__line"
      />
    </div>
  </div>
</DocumentFragment>
`;
